File: Question-2/library-System.py
class LibrarySystem:
    def __init__(self):
        self.books = {}
        self.reservations = {}

    def add_book(self, book_id, title):
        self.books[book_id] = {"title": title, "available": True, "reserved_by": None}

    def borrow_book(self, book_id, customer_id, is_digital=False):
        if book_id not in self.books:
            return "Book not found"
        book = self.books[book_id]
        if not book["available"]:
            if is_digital:
                return "Digital copy unavailable"
            return "Book is already borrowed. Added to the waitlist."
        book["available"] = False
        book["reserved_by"] = customer_id
        return f"{book['title']} borrowed by {customer_id} ({'Digital' if is_digital else 'Physical'})"

    def return_book(self, book_id):
        if book_id not in self.books:
            return "Book not found"
        book = self.books[book_id]
        book["available"] = True
        book["reserved_by"] = None
        return f"{book['title']} returned"

    def reserve_event(self, event_id, customer_id):
        if event_id not in self.reservations:
            self.reservations[event_id] = []
        if customer_id in self.reservations[event_id]:
            return "Already reserved"
        self.reservations[event_id].append(customer_id)
        return f"{customer_id} reserved for event {event_id}"

    def check_availability(self, book_id):
        return self.books.get(book_id, {}).get("available", False)
        
library = LibrarySystem()
library.add_book(1, "The handmaids tale")
print(library.borrow_book(1, "customer1", False))
print(library.check_availability(1))
print(library.return_book(1))
print(library.check_availability(1))
print(library.reserve_event("Event1", "customer1"))

File: Question-2/library_system.kt
fun main(){
    val library = LibrarySystem()
    library.addBook(1, "The Girl With Louding Voice")
    println(library.borrowBook(1, "Emebet", true))
    println(library.checkAvailability(1))
    println(library.returnBook(1))
    println(library.checkAvailability(1))
    println(library.reserveEvent("E1", "customer1"))

}
class LibrarySystem {
    val books = mutableMapOf<Int, Book>()
    val reservations = mutableMapOf<String, MutableList<String>>()
    data class Book(var title: String, var available: Boolean, var reservedBy: String?)
    fun addBook(bookId: Int, title: String) {
        books[bookId] = Book(title, true, null)
    }
    fun borrowBook(bookId: Int, customerID: String, isDigital: Boolean = false): String {
        val book = books[bookId] ?: return "Book not found"
        if (!book.available) {
            return if (isDigital) "Digital copy unavailable" else "Book is already borrowed. Added to the waitlist."
        }
        book.available = false
        book.reservedBy = customerID
        return "Book ${book.title} borrowed by $customerID (${if (isDigital) "Digital" else "Physical"})"
    }
    fun returnBook(bookId: Int): String {
        val book = books[bookId] ?: return "Book not found"
        book.available = true
        book.reservedBy = null
        return "Book ${book.title} returned"
    }
    fun reserveEvent(eventId: String, customerID: String): String {
        if (!reservations.containsKey(eventId)) {
            reservations[eventId] = mutableListOf()
        }
        val event = reservations[eventId]!!
        if (customerID in event) return "Already reserved"
        event.add(customerID)
        return "$customerID reserved for event $eventId"
    }
    fun checkAvailability(bookId: Int): Boolean {
        return books[bookId]?.available ?: false
    }
}